source/mqtt_client_connection.c

```diff
@@ -251,11 +272,115 @@ static void s_on_connection_resumed(
+                aws_napi_dispatch_threadsafe_function(
+                    env, binding->on_connection_success, NULL, on_success, num_params, params));
+        } else {
+            AWS_LOGF_INFO(
+                AWS_LS_NODEJS_CRT_GENERAL,
+                "Success callback invoked for connection binding %p but callback is null",
+                (void *)binding);
+        }
+    }
+    aws_mem_release(binding->allocator, args);
+}
+
+static void s_on_connection_success(
+    struct aws_mqtt_client_connection *connection,
+    enum aws_mqtt_connect_return_code return_code,
+    bool session_present,
+    void *user_data) {
+    (void)connection;
+
+    struct mqtt_connection_binding *binding = user_data;
+    binding->first_successfull_connection = true;
+    if (!binding->on_connection_success) {
+        return;
+    }
+
+    struct connection_success_args *args =
+        aws_mem_calloc(binding->allocator, 1, sizeof(struct connection_success_args));
+    AWS_FATAL_ASSERT(args);
+
+    args->return_code = return_code;
+    args->session_present = session_present;
+
+    AWS_NAPI_ENSURE(NULL, aws_napi_queue_threadsafe_function(binding->on_connection_success, args));
+}
+
+/*******************************************************************************
+ * on_connection_failure
+ ******************************************************************************/
+
+struct connection_failure_args {
+    int error_code;
+};
+static void s_on_connection_failure_call(napi_env env, napi_value on_failure, void *context, void *user_data) {
+    struct mqtt_connection_binding *binding = context;
+    struct connection_failure_args *args = user_data;
+    if (env) {
+        if (binding->on_connection_failure) {
+            napi_value params[1];
+            const size_t num_params = AWS_ARRAY_SIZE(params);
+
+            AWS_NAPI_ENSURE(env, napi_create_int32(env, args->error_code, &params[0]));
+
+            AWS_NAPI_ENSURE(
+                env,
+                aws_napi_dispatch_threadsafe_function(
+                    env, binding->on_connection_failure, NULL, on_failure, num_params, params));
+        } else {
+            AWS_LOGF_INFO(
+                AWS_LS_NODEJS_CRT_GENERAL,
+                "Failure callback invoked for connection binding %p but callback is null",
+                (void *)binding);
+        }
+    }
+    aws_mem_release(binding->allocator, args);
+}
+
+static void s_on_connection_failure(struct aws_mqtt_client_connection *connection, int error_code, void *user_data) {
+    (void)connection;
+
+    struct mqtt_connection_binding *binding = user_data;
+    if (!binding->on_connection_failure) {
+        return;
+    }
+
+    struct connection_failure_args *args =
+        aws_mem_calloc(binding->allocator, 1, sizeof(struct connection_failure_args));
+    AWS_FATAL_ASSERT(args);
+    args->error_code = error_code;
+
+    AWS_NAPI_ENSURE(NULL, aws_napi_queue_threadsafe_function(binding->on_connection_failure, args));
+}
```

Comment:
We probably also wanna release the function after callback is called.

Reply:
this is asynchronous, so it might be deleted before the callback goes through no?